pygfx/renderers/wgpu/__init__.py

```diff
@@ -1,11 +1,112 @@
+performed by the PipelineContainer. Some of these objects match 1-on-1
+to the world object, but others (shader module and pipeline) are
+specific to the environment. And then there are multiple blend-passes
+to create as well. The environment also plays a role in "finalizing"
+the final environment-specific objects.
+
+                     ┌────────────────┐  ┌──────────────────────────┐  ┌───────────────────────┐
+   RenderContext ───►│WGSL source     │  │BindGroupLayoutDescriptors│  │VertexLayoutDescriptors│
+   specific          └───────┬────────┘  └──────────┬───────────────┘  └──────────┬────────────┘
+   templating                │                      │                             │
+                     ┌───────▼────────┐  ┌──────────▼─────────┐                   │
+                     │WgpuShaderModule│  │WgpuBindGroupLayouts│                   │
+                     └───────┬────────┘  └──────────┬─────────┘                   │
+                             │                      │                             │
+                             └──────────────────┐   │  ┌──────────────────────────┘
+                                                │   │  │
+                                            ┌───▼───▼──▼───┐         ┌─────────────┐
+                                            │ WgpuPipeline ◄─────────┤pipeline info│
+                                            └───────┬──────┘         └─────────────┘
+                                                    │
+                          Ubuffers                  │
+                          Buffers   ────────────┐   │
+                          Textures              │   │
+                             │                  │   │
+                             │                  │   │
+                      ┌──────▼───────┐      ┌───▼───▼─────────┐      ┌───────────┐
+                      │WgpuBindGroups├──────► Dispatch (draw) ◄──────┤render info│
+                      └──────────────┘      └─────────────────┘      └───────────┘
+
+## Reacting to changes
+
+To keep a PipelineContainer up-to-date we need to do two things. First
+any changes in the world object and/or material must be applied. This
+is done via the Trackable classes, and affects the second level. Next,
+some of these changes invalidate the objects in level 3, so we need to
+detect that as well.
+
+We could keep a global list of world objects that need an update, but
+the fact that a world object can be used in multiple environments makes
+this complex. So instead a renderer iterates over all world objects,
+triggering updates as needed.
+
+TODO: how we keep buffers and textures up-to-date
+
+## Caching
+
+Objects are stored for each specific environment type, by using the
+environment's hash. The environment includes a system to detect that
+it is no longer used to that all objects related to that environment
+can be cleaned up.
+
+We also want to re-use wgpu objects like pipelines and shadermodules.
```

Comment:
To be clear: this is not supported yet, right?

Reply:
Nope. Made that more clear.